src/metrics-gatherer.ts

```diff
@@ -162,59 +183,63 @@ export class MetricsGatherer {
 	// check that a metric is of the given type(s)
 	public checkMetricType(name: string, kinds: string[]) {
 		try {
-			return kinds.includes(this.kinds[name]);
+			return kinds.includes(this.meta[name].kind);
 		} catch (e) {
 			this.err(e);
 		}
 	}
 
+	public exists(name: string) {
+		// ensure meta exists, and use meta to determine kind, ensuring
+		// prometheus.Metric object of that kind exists
+		return this.meta[name] && this.metrics[this.meta[name].kind][name];
+	}
+
 	// used declaratively to ensure a given metric of a certain kind exists,
 	// given some custom params to instantiate it if absent
-	public ensureExists(
+	private ensureExists(
+		kind: Kind,
 		name: string,
-		kind: string,
-		labels: LabelSet = {},
-		custom: CustomParams = {},
+		customParams: CustomParams = {},
 	) {
 		try {
 			// if exists, bail early
-			if (this.existMap[name]) {
+			if (this.exists(name)) {
 				return;
 			}
-			// if already exists with another kind, throw error
-			if (this.kinds[name] && this.kinds[name] !== kind) {
+			// if no meta, describe by default to satisfy prometheus
+			if (!this.meta[name]) {
+				this.describe[kind](name, `undescribed ${kind} metric`, customParams);
+			} else if (this.meta[name].kind !== kind) {
+				// if name already associated with another kind, throw error
 				throw new MetricsGathererError(
 					`tried to use ${name} twice - first as ` +
-						`${this.kinds[name]}, then as ${kind}`,
+						`${this.meta[name].kind}, then as ${kind}`,
 				);
 			}
-			// if not described, describe (poorly)
-			if (!(name in this.descriptions)) {
-				this.descriptions[name] = `undescribed ${kind} metric`;
-			}
-			// if doesn't exist, create metric
+			// create prometheus.Metric instance
 			this.metrics[kind][name] = constructors[kind].create({
 				name,
-				help: this.descriptions[name],
-				labelNames: Object.keys(labels),
-				...custom,
-				...this.customParams[name],
+				help: this.meta[name].help,
+				...customParams,
+				...this.meta[name].customParams,
 			});
-			this.kinds[name] = kind;
-			this.existMap[name] = true;
 		} catch (e) {
 			this.err(e);
 		}
 	}
 
+	public getMetric(name: string): prometheus.Metric | undefined {
+		return this.exists(name)
+			? this.metrics[this.meta[name].kind][name]
+			: undefined;
+	}
+
 	// reset a given metric by name
 	public reset(name: string) {
 		try {
-			// will only have an entry in this.kinds if the metric has been observed
-			// at least once (don't need to "reset" otherwise, and avoids error in
-			// indexing)
-			if (this.kinds[name]) {
-				this.metrics[this.kinds[name]][name].reset();
+			if (this.exists(name)) {
```

Comment:
This can be
```suggestion
			const metric = this.exists(name)
			if (metric ) {
				metric.reset() // No need for the `!` cast anymore because typescript can see it's not null
			}
```
because `this.exists` actually returns the metric and removes the need for `getMetric` to duplicate that behaviour.  that said I think it would make sense to actually do:
```ts
	public getMetric(name: string): prometheus.Metric | undefined {
		if(this.meta[name]) {
			return this.metrics[this.meta[name].kind][name];
		}
	}
	public exists(name: string) {
		return this.getMetrics(name) != null;
	}
```
```suggestion
			const metric = this.getMetric(name)
			if (metric ) {
				metric.reset() // No need for the `!` cast anymore because typescript can see it's not null
			}
```
so that `exists` returns a straight boolean, but makes use of `getMetric` to avoid duplication and double fetching of the metric

Reply:
a very clear improvement, cheers!